[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34234 [WIP] Changes in test/test_jit_fuser.py test/test_jit_fuser_te.py.
* #34233 [WIP] Changes to peephole.cpp.
* #34231 [WIP] Guard elimination changes from bertmaher/pytorch_fusion.
* #34230 [TensorExpr] Add tensorexpr benchmarks.
* #34228 [TensorExpr] Add LLVM codegen.
* #34227 [TensorExpr] Add CUDA codegen.
* #34226 [TensorExpr] Add a fuser pass based on tensor expressions.
* **#34224 [TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch.**

Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

Commit:
Move tensor expressions work from bertmaher/pytorch to pytorch/pytorch.

This PR moves changes to tensor expressions IR from from
bertmaher/pytorch_fusion. The expressions are not used anywhere yet,
only in tests.

[ghstack-poisoned]

Commit:
Update on "Move tensor expressions work from bertmaher/pytorch to pytorch/pytorch."


This PR moves changes to tensor expressions IR from from
bertmaher/pytorch_fusion. The expressions are not used anywhere yet,
only in tests.

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Update descriptions. on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Lints on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Accepting as it's been reviewed on the branch and changes are local to tensorexpr folder only

Commit:
Rebase on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

bertmaher reviewed (approved):
Reviewed in the side branch.  I'm going to stamp this so we can move our development to master without any more monolithic commits.

Commit:
Rebase and try to fix windows build. on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Pull changes to core classes for representing expressions and statements from the side branch."


Our development has been happening on a side branch `pytorch_fusion` in
`bertmaher/pytorch` fork. This PR moves changes to the core classes
representing expressions and transformations on them.

At this moment, the tensor expressions are only used in tests.
Subsequent PRs add LLVM and CUDA codegen for tensor expressions and
implement fuser on top of these.

This PR is huge as it is a squashed version of changes in the side
branch. It is not practical to pull changes one by one from the branch,
so here is the squashed version. If you're interested in seeing the
history of changes, please refer to https://github.com/bertmaher/pytorch

Differential Revision: [D20251835](https://our.internmc.facebook.com/intern/diff/D20251835)

[ghstack-poisoned]
